
<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="This contains all material for Markdown version of BTP workshops.">
      
      
      
        <meta name="author" content="Authors">
      
      
        <link rel="shortcut icon" href="../../../assets/images/favicon.png">
      
      <meta name="generator" content="mkdocs-0.16.3, mkdocs-material-1.7.3">
    
    
      
        <title>Scaffolding - Markdown Manuals for Bioinformatics Training Platform (BTP)</title>
      
    
    
      <script src="../../../assets/javascripts/modernizr-1df76c4e58.js"></script>
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/application-acc56dad49.css">
      
        <link rel="stylesheet" href="../../../assets/stylesheets/application-8817cfa535.palette.css">
      
    
    
      
        
        
        
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../../../css/details.css">
    
    
  </head>
  
  
  
  
    <body data-md-color-primary="indigo" data-md-color-accent="light-blue">
  
    <svg class="md-svg">
      <defs>
        
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <a href="../../.." title="Markdown Manuals for Bioinformatics Training Platform (BTP)" class="md-logo md-header-nav__button">
            <img src="../../../media/logos/bpalogo.png" width="24" height="24">
          </a>
        
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <span class="md-flex__ellipsis md-header-nav__title">
          
            
              
                <span class="md-header-nav__parent">
                  DeNovo Modules
                </span>
              
            
            Scaffolding
          
        </span>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
          
<div class="md-search" data-md-component="search">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" required placeholder="Search" accesskey="s" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset">close</button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result" data-md-lang-search="">
          <div class="md-search-result__meta" data-md-lang-result-none="No matching documents" data-md-lang-result-one="1 matching document" data-md-lang-result-other="# matching documents">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
      
        

  

<nav class="md-tabs md-tabs--active" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  <li class="md-tabs__item">
    
      <a href="../../.." title="Home" class="md-tabs__link">
        Home
      </a>
    
  </li>

      
        
      
        
  
  
    <li class="md-tabs__item">
      
        <a href="../../../timetables/timetable_introNGS/" title="Timetable" class="md-tabs__link">
          Timetable
        </a>
      
    </li>
  

      
        
      
        
  
  
    <li class="md-tabs__item">
      
        <a href="../../btp-module-ngs-cli/commandline/" title="Introduction to NGS Modules" class="md-tabs__link">
          Introduction to NGS Modules
        </a>
      
    </li>
  

      
        
  
  
    <li class="md-tabs__item">
      
        <a href="../../cancer-module-alignment/alignment/" title="Cancer Genomics Modules" class="md-tabs__link">
          Cancer Genomics Modules
        </a>
      
    </li>
  

      
        
  
  
    <li class="md-tabs__item">
      
        <a href="../../denovo-module-bac/bac_asm/" title="DeNovo Modules" class="md-tabs__link md-tabs__link--active">
          DeNovo Modules
        </a>
      
    </li>
  

      
        
  
  
    <li class="md-tabs__item">
      
        <a href="../../metagenomics-module-qc/ngs-qc/" title="Metagenomics Modules" class="md-tabs__link">
          Metagenomics Modules
        </a>
      
    </li>
  

      
    </ul>
  </div>
</nav>
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    
      <i class="md-logo md-nav__button">
        <img src="../../../media/logos/bpalogo.png">
      </i>
    
    Markdown Manuals for Bioinformatics Training Platform (BTP)
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../trainers/trainers/" title="The Trainers" class="md-nav__link">
      The Trainers
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3">
    
    <label class="md-nav__link" for="nav-3">
      Timetable
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Timetable
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../../timetables/timetable_introNGS/" title="Intro to NGS" class="md-nav__link">
      Intro to NGS
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../../timetables/timetable_cancer/" title="Cancer Genomics" class="md-nav__link">
      Cancer Genomics
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../../timetables/timetable_longRead/" title="Long Read" class="md-nav__link">
      Long Read
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../preamble/" title="Workshop Information" class="md-nav__link">
      Workshop Information
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5">
    
    <label class="md-nav__link" for="nav-5">
      Introduction to NGS Modules
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-5">
        Introduction to NGS Modules
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-ngs-cli/commandline/" title="Introduction to Command Line" class="md-nav__link">
      Introduction to Command Line
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-ngs-qc/ngs-qc/" title="Data Quality" class="md-nav__link">
      Data Quality
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-ngs-mapping/alignment/" title="Read Alignment" class="md-nav__link">
      Read Alignment
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-chip-seq/chip-seq/" title="ChIP-seq" class="md-nav__link">
      ChIP-seq
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-rna-seq/rna-seq/" title="RNA-Seq" class="md-nav__link">
      RNA-Seq
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-ngs-bio/biol/" title="NGS Biological Insight" class="md-nav__link">
      NGS Biological Insight
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-velvet/velvet/" title="de novo Genome Assembly" class="md-nav__link">
      de novo Genome Assembly
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../btp-module-denovo-canu/denovo_canu/" title="DeNovo Canu" class="md-nav__link">
      DeNovo Canu
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6">
    
    <label class="md-nav__link" for="nav-6">
      Cancer Genomics Modules
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        Cancer Genomics Modules
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../cancer-module-alignment/alignment/" title="Read Alignment" class="md-nav__link">
      Read Alignment
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../cancer-module-snv/snv/" title="Single Nucleotide Variant Calling and Annotation" class="md-nav__link">
      Single Nucleotide Variant Calling and Annotation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../cancer-module-cnv/cnv-tut/" title="Copy Number Variation" class="md-nav__link">
      Copy Number Variation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../cancer-module-sv/sv_tut/" title="Structural Variant Analysis" class="md-nav__link">
      Structural Variant Analysis
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../cancer-module-viz/visu/" title="Variant Visualisation" class="md-nav__link">
      Variant Visualisation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6-6" type="checkbox" id="nav-6-6">
    
    <label class="md-nav__link" for="nav-6-6">
      Somatic Mutations
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="2">
      <label class="md-nav__title" for="nav-6-6">
        Somatic Mutations
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../cancer-module-somatic/01_signatures/" title="Assessing somatic mutational signatures" class="md-nav__link">
      Assessing somatic mutational signatures
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../cancer-module-somatic/02_intogen/" title="Cohort analysis for the identification of driver genes" class="md-nav__link">
      Cohort analysis for the identification of driver genes
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7" checked>
    
    <label class="md-nav__link" for="nav-7">
      DeNovo Modules
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-7">
        DeNovo Modules
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../denovo-module-bac/bac_asm/" title="Bacterial Assembly" class="md-nav__link">
      Bacterial Assembly
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../denovo-module-euk/eukaryotic_assem/" title="Eukaryote Assembly" class="md-nav__link">
      Eukaryote Assembly
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        Scaffolding
      </label>
    
    <a href="./" title="Scaffolding" class="md-nav__link md-nav__link--active">
      Scaffolding
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#key-learning-outcomes" title="Key Learning Outcomes" class="md-nav__link">
    Key Learning Outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources-youll-be-using" title="Resources You’ll be Using" class="md-nav__link">
    Resources You’ll be Using
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tools-used" title="Tools Used" class="md-nav__link">
    Tools Used
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#author-information" title="Author Information" class="md-nav__link">
    Author Information
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scaffolding-with-long-mate-pair-libraries-chalara-scaffolding-using-lmp" title="Scaffolding with long mate pair libraries: Chalara scaffolding using LMP" class="md-nav__link">
    Scaffolding with long mate pair libraries: Chalara scaffolding using LMP
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pair-end-assembly-using-chalara-dataset" title="Pair-end assembly using Chalara dataset" class="md-nav__link">
    Pair-end assembly using Chalara dataset
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chalara-scaffolding-using-long-mate-pair-reads-lmp" title="Chalara scaffolding using long mate-pair reads (LMP)" class="md-nav__link">
    Chalara scaffolding using long mate-pair reads (LMP)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chalara-beyond-first-pass" title="Chalara: beyond first pass" class="md-nav__link">
    Chalara: beyond first pass
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" title="References" class="md-nav__link">
    References
  </a>
  
</li>
      
      
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-8" type="checkbox" id="nav-8">
    
    <label class="md-nav__link" for="nav-8">
      Metagenomics Modules
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-8">
        Metagenomics Modules
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../metagenomics-module-qc/ngs-qc/" title="Metagenomics QC" class="md-nav__link">
      Metagenomics QC
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../metagenomics-module-tax/tax/" title="Metagenomics Taxonomic" class="md-nav__link">
      Metagenomics Taxonomic
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../metagenomics-module-vis/vis/" title="Metagenomics Visualisation" class="md-nav__link">
      Metagenomics Visualisation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../metagenomics-module-wga/wga/" title="Metagenomics WGA" class="md-nav__link">
      Metagenomics WGA
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../metagenomics-module-fda/fda/" title="Metagenomics Functional Analysis" class="md-nav__link">
      Metagenomics Functional Analysis
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#key-learning-outcomes" title="Key Learning Outcomes" class="md-nav__link">
    Key Learning Outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources-youll-be-using" title="Resources You’ll be Using" class="md-nav__link">
    Resources You’ll be Using
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tools-used" title="Tools Used" class="md-nav__link">
    Tools Used
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#author-information" title="Author Information" class="md-nav__link">
    Author Information
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scaffolding-with-long-mate-pair-libraries-chalara-scaffolding-using-lmp" title="Scaffolding with long mate pair libraries: Chalara scaffolding using LMP" class="md-nav__link">
    Scaffolding with long mate pair libraries: Chalara scaffolding using LMP
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pair-end-assembly-using-chalara-dataset" title="Pair-end assembly using Chalara dataset" class="md-nav__link">
    Pair-end assembly using Chalara dataset
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chalara-scaffolding-using-long-mate-pair-reads-lmp" title="Chalara scaffolding using long mate-pair reads (LMP)" class="md-nav__link">
    Chalara scaffolding using long mate-pair reads (LMP)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chalara-beyond-first-pass" title="Chalara: beyond first pass" class="md-nav__link">
    Chalara: beyond first pass
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" title="References" class="md-nav__link">
    References
  </a>
  
</li>
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                  <h1>Scaffolding</h1>
                
                <h2 id="key-learning-outcomes">Key Learning Outcomes<a class="headerlink" href="#key-learning-outcomes" title="Permanent link">&para;</a></h2>
<p>After completing this module the trainee should be able to:</p>
<ul>
<li>
<p>Run an assembly using long mate pair (LMP) reads.</p>
</li>
<li>
<p>Compare results of LMP assemblies with those from the short reads.</p>
</li>
<li>
<p>Run scaffolding and assess the assembly statistics.</p>
</li>
</ul>
<hr />
<h2 id="resources-youll-be-using">Resources You’ll be Using<a class="headerlink" href="#resources-youll-be-using" title="Permanent link">&para;</a></h2>
<h3 id="tools-used">Tools Used<a class="headerlink" href="#tools-used" title="Permanent link">&para;</a></h3>
<p>Kmer Analysis Tool kit:<br />
<a href="https://github.com/TGAC/KAT">https://github.com/TGAC/KAT</a></p>
<p>Nextclip:<br />
<a href="https://github.com/richardmleggett/nextclip">https://github.com/richardmleggett/nextclip</a></p>
<p>Abyss:<br />
<a href="http://www.bcgsc.ca/platform/bioinfo/software/abyss">http://www.bcgsc.ca/platform/bioinfo/software/abyss</a></p>
<p>Soap Denovo:<br />
<a href="http://soap.genomics.org.cn/soapdenovo.html">http://soap.genomics.org.cn/soapdenovo.html</a></p>
<p>SOAPec:<br />
<a href="http://soap.genomics.org.cn/about.html">http://soap.genomics.org.cn/about.html</a></p>
<hr />
<h2 id="author-information">Author Information<a class="headerlink" href="#author-information" title="Permanent link">&para;</a></h2>
<p><em>Primary Author(s):</em><br />
Bernardo Clavijo, TGAC <a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#66;&#101;&#114;&#110;&#97;&#114;&#100;&#111;&#46;&#67;&#108;&#97;&#118;&#105;&#106;&#111;&#64;&#116;&#103;&#97;&#99;&#46;&#97;&#99;&#46;&#117;&#107;">&#66;&#101;&#114;&#110;&#97;&#114;&#100;&#111;&#46;&#67;&#108;&#97;&#118;&#105;&#106;&#111;&#64;&#116;&#103;&#97;&#99;&#46;&#97;&#99;&#46;&#117;&#107;</a></p>
<p><em>Contributor(s):</em><br />
Sonika Tyagi, AGRF <a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#111;&#110;&#105;&#107;&#97;&#46;&#116;&#121;&#97;&#103;&#105;&#64;&#97;&#103;&#114;&#102;&#46;&#111;&#114;&#103;&#46;&#97;&#117;">&#115;&#111;&#110;&#105;&#107;&#97;&#46;&#116;&#121;&#97;&#103;&#105;&#64;&#97;&#103;&#114;&#102;&#46;&#111;&#114;&#103;&#46;&#97;&#117;</a><br />
Zhiliang Chan <a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#122;&#104;&#105;&#108;&#105;&#97;&#110;&#103;&#64;&#117;&#110;&#115;&#119;&#46;&#101;&#100;&#117;&#46;&#97;&#117;">&#122;&#104;&#105;&#108;&#105;&#97;&#110;&#103;&#64;&#117;&#110;&#115;&#119;&#46;&#101;&#100;&#117;&#46;&#97;&#117;</a></p>
<hr />
<h2 id="scaffolding-with-long-mate-pair-libraries-chalara-scaffolding-using-lmp">Scaffolding with long mate pair libraries: Chalara scaffolding using LMP<a class="headerlink" href="#scaffolding-with-long-mate-pair-libraries-chalara-scaffolding-using-lmp" title="Permanent link">&para;</a></h2>
<p>Scaffold is made of two or more contigs joined together using the read
pair information. In the absence of a high-quality reference genome, new
genome assemblies are often evaluated on the basis of the number of
scaffolds and contigs required to represent the genome. Other criteria
such as the alignment of reads back to assemblies, N50, and the length
of contigs and scaffolds relative to the size of the genome can also be
used to assess the quality of assemblies. In this exercise we will be
using a long mate pair data to perform assembly and scaffolding and we
will focus on how using LMP reads can affect the assemblies. Most part
of this tutorial has been precomputed and made available to you in
interest of time. We will spend more time exploring and discussing the
results.</p>
<h3 id="pair-end-assembly-using-chalara-dataset">Pair-end assembly using <em>Chalara</em> dataset<a class="headerlink" href="#pair-end-assembly-using-chalara-dataset" title="Permanent link">&para;</a></h3>
<p>Before doing a scaffolding, we need to build an assembly using the
pair-end reads first.</p>
<p>Let&rsquo;s go to the correct location where the files are:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>cd /home/trainee/scaffolding/cha_raw/test_1
</pre></div>
</td></tr></table>

<p><br></p>
<div class="admonition failure">
<p class="admonition-title">STOP</p>
<p>You DO NOT need to run this command. This has already been done for you.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>abyss-pe name=cha1 k=27 in=&quot;../LIB2570_raw_R1.fastq ../LIB2570_raw_R2.fastq&quot; np=4
</pre></div>
</td></tr></table>

</div>
<p><br>
Let’s look at the stats by doing:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>less cha1-stats.tab
</pre></div>
</td></tr></table>

<p><br>
<strong>Table 1</strong>: Statistics of <em>Chalara</em> assembly by ABySS using k=27.</p>
<table>
<thead>
<tr>
<th><strong>n</strong></th>
<th><strong>n:500</strong></th>
<th><strong>L50</strong></th>
<th><strong>min</strong></th>
<th><strong>N80</strong></th>
<th><strong>N50</strong></th>
<th><strong>N20</strong></th>
<th><strong>E-size</strong></th>
<th><strong>max</strong></th>
<th><strong>sum</strong></th>
<th><strong>name</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>394789</td>
<td>11681</td>
<td>2094</td>
<td>500</td>
<td>2880</td>
<td>6254</td>
<td>12303</td>
<td>7936</td>
<td>44414</td>
<td>44.07e6</td>
<td>cha1-unitigs.fa</td>
</tr>
<tr>
<td>394199</td>
<td>11673</td>
<td>2097</td>
<td>500</td>
<td>2887</td>
<td>6255</td>
<td>12303</td>
<td>7937</td>
<td>44414</td>
<td>44.11e6</td>
<td>cha1-contigs.fa</td>
</tr>
<tr>
<td>394161</td>
<td>11647</td>
<td>2095</td>
<td>500</td>
<td>2898</td>
<td>6269</td>
<td>12303</td>
<td>7944</td>
<td>44414</td>
<td>44.12e6</td>
<td>cha1-scaffolds.fa</td>
</tr>
</tbody>
</table>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>cd /home/trainee/scaffolding/cha_raw/test_2
</pre></div>
</td></tr></table>

<p><br></p>
<div class="admonition failure">
<p class="admonition-title">STOP</p>
<p>This is also a pre-computed example for you:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span> abyss-pe name=cha2 k=61 in=&quot;../LIB2570_raw_R1.fastq ../LIB2570_raw_R2.fastq&quot; np=4
</pre></div>
</td></tr></table>

</div>
<p><br>
Again, we need to look at the statistics of our new assembly using k=61:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>less cha2-stats.tab
</pre></div>
</td></tr></table>

<p><br>
It should be looking like this:</p>
<p><strong>Table 2</strong>: Statistics of <em>Chalara</em> assembly by ABySS using k=61.</p>
<table>
<thead>
<tr>
<th><strong>n</strong></th>
<th><strong>n:500</strong></th>
<th><strong>L50</strong></th>
<th><strong>min</strong></th>
<th><strong>N80</strong></th>
<th><strong>N50</strong></th>
<th><strong>N20</strong></th>
<th><strong>E-size</strong></th>
<th><strong>max</strong></th>
<th><strong>sum</strong></th>
<th><strong>name</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>130547</td>
<td>12596</td>
<td>1770</td>
<td>500</td>
<td>3352</td>
<td>8379</td>
<td>16380</td>
<td>10518</td>
<td>54300</td>
<td>50.75e6</td>
<td>cha2-unitigs.fa</td>
</tr>
<tr>
<td>130210</td>
<td>12575</td>
<td>1771</td>
<td>500</td>
<td>3363</td>
<td>8382</td>
<td>16380</td>
<td>10525</td>
<td>54300</td>
<td>50.78e6</td>
<td>cha2-contigs.fa</td>
</tr>
<tr>
<td>130182</td>
<td>12555</td>
<td>1769</td>
<td>500</td>
<td>3377</td>
<td>8394</td>
<td>16380</td>
<td>10534</td>
<td>54300</td>
<td>50.78e6</td>
<td>cha2-scaffolds.fa</td>
</tr>
</tbody>
</table>
<p><br></p>
<div class="admonition note">
<p class="admonition-title">Question</p>
<p>The assembly contiguity is worse than <em>Fusarium</em>, why?</p>
<div class="admonition success"><details><summary><strong>Answer</strong></summary><ul><li>Genome characteristics.</li><li>Data</li><li>Coverage</li><li>Error distributions</li><li>Read sizes</li><li>Fragment sizes</li><li>Kmer spectra</li></ul></details></div>
</div>
<h3 id="chalara-scaffolding-using-long-mate-pair-reads-lmp">Chalara scaffolding using long mate-pair reads (LMP)<a class="headerlink" href="#chalara-scaffolding-using-long-mate-pair-reads-lmp" title="Permanent link">&para;</a></h3>
<p>Let’s put some LMP in there.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>cd /home/trainee/scaffolding/cha_raw/test_3
</pre></div>
</td></tr></table>

<p><br></p>
<div class="admonition failure">
<p class="admonition-title">STOP</p>
<p>This is also a pre-computed example for you:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>abyss-pe name=chalmp1 k=61 in=&quot;../LIB2570_raw_R1.fastq ../LIB2570_raw_R2.fastq&quot; mp=&quot;lmp1&quot; lmp1=&quot;../LIB8209_raw_R1.fastq ../LIB8209_raw_R2.fastq&quot; np=4
</pre></div>
</td></tr></table>

</div>
<p><br>
  <table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>less chalmp1-stats.tab
</pre></div>
</td></tr></table></p>
<p><br>
<strong>Table 3</strong>: Statistics of <em>Chalara</em> assembly by ABySS using k=27.</p>
<table>
<thead>
<tr>
<th><strong>n</strong></th>
<th><strong>n:500</strong></th>
<th><strong>L50</strong></th>
<th><strong>min</strong></th>
<th><strong>N80</strong></th>
<th><strong>N50</strong></th>
<th><strong>N20</strong></th>
<th><strong>E-size</strong></th>
<th><strong>max</strong></th>
<th><strong>sum</strong></th>
<th><strong>name</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>130548</td>
<td>12596</td>
<td>1770</td>
<td>500</td>
<td>3352</td>
<td>8379</td>
<td>16380</td>
<td>10518</td>
<td>54300</td>
<td>50.75e6</td>
<td>chalmp1-unitigs.fa</td>
</tr>
<tr>
<td>130211</td>
<td>12575</td>
<td>1771</td>
<td>500</td>
<td>3363</td>
<td>8382</td>
<td>16380</td>
<td>10525</td>
<td>54300</td>
<td>50.78e6</td>
<td>chalmp1-contigs.fa</td>
</tr>
<tr>
<td>130148</td>
<td>12545</td>
<td>1769</td>
<td>500</td>
<td>3380</td>
<td>8400</td>
<td>16380</td>
<td>10535</td>
<td>54300</td>
<td>50.78e6</td>
<td>chalmp1-scaffolds.fa</td>
</tr>
</tbody>
</table>
<p>So&hellip; what happened? Let’s check the data by:</p>
<ul>
<li>Kmer spectra</li>
<li>Fragment sizes</li>
</ul>
<p><br>
Any hints on the protocol? And a not-so-obvious property:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>kat plot spectra-mx --intersection -x 30 -y 15000000 -o pe_vs_lmp-main.mx.spectra-mx.png pe_vs_lmp-main.mx
</pre></div>
</td></tr></table>

<p><br>
Do you remember LMP required pre-processing? Let’s try with processed LMP.</p>
<p>Prior task (already made) preprocess the LMP with <code class="codehilite">nextclip</code>.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>cd /home/trainee/scaffolding/cha_proc/test_1
</pre></div>
</td></tr></table>

<p><br></p>
<div class="admonition failure">
<p class="admonition-title">STOP</p>
<p>This is also a pre-computed example for you:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>abyss-pe name=chalmpproc1 k=61 in=&quot;../../cha_raw/LIB2570_raw_R1.fastq ../../cha_raw/LIB2570_raw_R2.fastq&quot; mp=&quot;proclmp1&quot; proclmp1=&quot;../LIB8209_preproc_R1.fastq ../LIB8209_preproc_R2.fastq&quot; np=4
</pre></div>
</td></tr></table>

</div>
<p><br></p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>less chalmpproc1-stats.tab
</pre></div>
</td></tr></table>

<p><br>
<strong>Table 4</strong>: Statistics of <em>Chalara</em> assembly by ABySS using k=61 with LMP.</p>
<table>
<thead>
<tr>
<th><strong>n</strong></th>
<th><strong>n:500</strong></th>
<th><strong>L50</strong></th>
<th><strong>min</strong></th>
<th><strong>N80</strong></th>
<th><strong>N50</strong></th>
<th><strong>N20</strong></th>
<th><strong>E-size</strong></th>
<th><strong>max</strong></th>
<th><strong>sum</strong></th>
<th><strong>name</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>130548</td>
<td>12596</td>
<td>1770</td>
<td>500</td>
<td>3352</td>
<td>8379</td>
<td>16380</td>
<td>10518</td>
<td>54300</td>
<td>50.75e6</td>
<td>chalmpproc1-unitigs.fa</td>
</tr>
<tr>
<td>130211</td>
<td>12575</td>
<td>1771</td>
<td>500</td>
<td>3363</td>
<td>8382</td>
<td>16380</td>
<td>10525</td>
<td>54300</td>
<td>50.78e6</td>
<td>chalmpproc1-contigs.fa</td>
</tr>
<tr>
<td>122061</td>
<td>6679</td>
<td>167</td>
<td>500</td>
<td>18609</td>
<td>87510</td>
<td>187171</td>
<td>106178</td>
<td>397967</td>
<td>51.15e6</td>
<td>chalmpproc1-scaffolds.fa</td>
</tr>
</tbody>
</table>
<p><br>
That’s much better!</p>
<hr />
<h2 id="chalara-beyond-first-pass">Chalara: beyond first pass<a class="headerlink" href="#chalara-beyond-first-pass" title="Permanent link">&para;</a></h2>
<p>Do you remember these?</p>
<p>Genome characteristics. Data:</p>
<ul>
<li>Coverage</li>
<li>Error distributions</li>
<li>Read sizes</li>
<li>Fragment sizes</li>
<li>Kmer spectra</li>
</ul>
<p>Let’s think a bit and try to improve the assembly.</p>
<div class="admonition note">
<p class="admonition-title">Question</p>
<p>If you look at the pre-processed LMP, do you notice anything peculiar?</p>
</div>
<p>Testing the inclusion of heavily pre-processed LMP coverage into the DBG</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>cd /home/trainee/scaffolding/cha_proc/test_2
</pre></div>
</td></tr></table>

<p><br></p>
<div class="admonition failure">
<p class="admonition-title">STOP</p>
<p>This is also a pre-computed example for you:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>abyss-pe name=chalmp2 k=61 se=&quot;../LIB8209_preproc_single.fastq&quot; lib=&quot;lmp1&quot; lmp1=&quot;../LIB8209_preproc_R1.fastq ../LIB8209_preproc_R2.fastq&quot; np=4 &gt;chaproclmp2.log 2&gt;&amp;1
</pre></div>
</td></tr></table>

</div>
<p><br></p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>less chalmp2-stats.tab
</pre></div>
</td></tr></table>

<p><br>
Let’s look at the stats.  </p>
<p><strong>Table 5</strong>: Statistics of improved <em>Chalara</em> assembly by ABySS using k=61 with LMP.</p>
<table>
<thead>
<tr>
<th><strong>n</strong></th>
<th><strong>n:500</strong></th>
<th><strong>L50</strong></th>
<th><strong>min</strong></th>
<th><strong>N80</strong></th>
<th><strong>N50</strong></th>
<th><strong>N20</strong></th>
<th><strong>E-size</strong></th>
<th><strong>max</strong></th>
<th><strong>sum</strong></th>
<th><strong>name</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>128306</td>
<td>10150</td>
<td>1182</td>
<td>500</td>
<td>4954</td>
<td>12585</td>
<td>24777</td>
<td>15693</td>
<td>68684</td>
<td>51.03e6</td>
<td>chaproclmp4-unitigs.fa</td>
</tr>
<tr>
<td>118939</td>
<td>5772</td>
<td>362</td>
<td>500</td>
<td>15717</td>
<td>41870</td>
<td>82033</td>
<td>52819</td>
<td>252066</td>
<td>52.24e6</td>
<td>chaproclmp4-contigs.fa</td>
</tr>
<tr>
<td>116139</td>
<td>4014</td>
<td>141</td>
<td>500</td>
<td>41145</td>
<td>109273</td>
<td>224986</td>
<td>133450</td>
<td>460979</td>
<td>52.56e6</td>
<td>chaproclmp4-scaffolds.fa</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="references">References<a class="headerlink" href="#references" title="Permanent link">&para;</a></h2>
<ol>
<li>
<p>Robert Ekblom* and Jochen B. W. Wolf. &ldquo;A field guide to whole-genome
    sequencing, assembly and annotation&rdquo;. Evolutionary Applications
    Special Issue: Evolutionary Conservation Volume 7, Issue 9, pages
    1026 – 1042, November 2014</p>
</li>
<li>
<p>De novo genome assembly: what every biologist should know. Nature
    Methods 9, 333 – 337 (2012) doi:10.1038/nmeth.1935</p>
</li>
</ol>
                
                  
                
              
              
                
              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../../denovo-module-euk/eukaryotic_assem/" title="Eukaryote Assembly" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Eukaryote Assembly
              </span>
            </div>
          </a>
        
        
          <a href="../../metagenomics-module-qc/ngs-qc/" title="Metagenomics QC" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Metagenomics QC
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            &copy; The authors; 2017-
          </div>
        
        powered by
        <a href="http://www.mkdocs.org" title="MkDocs">MkDocs</a>
        and
        <a href="http://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../../assets/javascripts/application-b30566c560.js"></script>
      
      
      <script>app.initialize({url:{base:"../../.."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../../js/details.js"></script>
      
    
    
      
    
  </body>
</html>